dlib/dnn/trainer.h

```diff
@@ -1012,9 +1012,28 @@ namespace dlib
                 // previously saved state in the hopes that the problem won't reoccur.
                 if (loss_increased_since_last_disk_sync()) 
                 {
+                    // keep the loss values, because otherwise we might end up never shrinking the learning rate
+                    const auto previous_loss_values_to_keep = std::move(previous_loss_values);
+                    const auto test_previous_loss_values_to_keep = std::move(test_previous_loss_values);
+
+                    // keep also the step counts, in order to avoid confusion
+                    const auto train_one_step_calls_to_keep = train_one_step_calls;
+                    const auto test_one_step_calls_to_keep = test_one_step_calls;
+                    const unsigned long steps_without_progress_to_keep = steps_without_progress;
+                    const unsigned long test_steps_without_progress_to_keep = test_steps_without_progress;
+
```

Comment:
I guess these 6 lines of code could be moved to a function that returns a struct. Then, after deserializing, the struct could be passed to another function.

At only 6 lines (*), not sure if doing so would make things any cleaner. But if there will be more variables that should be handled like this, then it'd make sense, I suppose.

(*): Or even less? Still not sure about the step counts...